### PR TITLE
feat(kernels): Triton FlashAttention LSE export + varlen packing

### PR DESCRIPTION
**Pull request by Billy1900**
## Summary

First step of a stack toward "Fused FlashAttention with causal mask, varlen
packing, and exported attention LSE" (see discussion in the repo). Rather than
attempting SM90 WGMMA+TMA + SM80 + ROCm MFMA + Triton all in one PR, this
lands the Triton piece first: it's the only backend where a real kernel
already existed to extend (`rl_engine/kernels/ops/triton/triton_attn.py`), and
it becomes the cross-platform semantic baseline the SM90/SM80/ROCm kernels
get checked against later, rather than against each other.

## What this adds

- **Attention-domain LSE export**: `return_lse=True` on the existing dense
  `triton_flash_attention`. No new kernel math — `M` (running max) and `L`
  (running sum-exp) were already computed online for the backward pass, just
  never returned. LSE is marked non-differentiable (`ctx.mark_non_differentiable`),
  matching `flash_attn`'s external `softmax_lse` contract. Default calls
  (`return_lse=False`) are unchanged — no breaking change to the (currently
  uncalled) public API.
- **`triton_flash_attention_varlen`**: a new packed variable-length path over
  `cu_seqlens`-style `[total_tokens, H, D]` tensors (same convention as
  `flash_attn_varlen_func` and this repo's `pack` op, #182), with forward +
  backward kernels. Causal masking uses the `Skv - Sq` anchor from
  `NativeAttentionOp` (docs/operators/attention.md) so prefill (`Sq == Skv`)
  and decode (`Sq < Skv`) share one formula.

**Explicitly out of scope for this PR**: SM90 WGMMA+TMA, SM80 `mma.sync`, and
ROCm MFMA kernels, and GQA support (the pre-existing dense Triton kernel
didn't support GQA either — not a regression). Not wired into
`KernelRegistry` yet.

## Correctness

Both paths are checked against an **independent** fp32 masked-softmax +
logsumexp reference (not against each other, so a shared bug in the
online-softmax loop can't hide). Run end-to-end on real H100 SXM5 hardware:

- Dense: `out` max-abs-diff ≈ `1.4e-3`, `lse` ≈ `1e-6` vs. reference (fp16
  inputs, fp32 accumulation).
- Varlen: forward + backward (`dq`/`dk`/`dv`) checked across causal/non-causal,
  decode-style (`Sq < Skv`, varying independently per sequence in the batch),
  sequence lengths deliberately **not** aligned to the kernel's 64-token block
  size, `head_dim` 64 and 128, and a zero-length sequence in the batch (a real
  case for fully-masked/empty RL responses — packed tensors have the *next*
  sequence's data immediately after a short one, so this is a genuine
  correctness boundary, not just an edge case to tolerate).

8/8 tests in `tests/test_triton_attention_varlen.py` pass; the pre-existing
26/26 in `tests/test_attention.py` are untouched and still pass.

## Docs

Added `docs/operators/attention-varlen.md` (template-following operator page,
wired into `docs/.nav.yml` and `docs/operators/README.md`), including an
explicit note that this LSE is attention-domain, not the vocab-domain LSE the
`fused_logp`/`linear_logp` kernels already use internally — same name,
different tensor, worth flagging to avoid confusion.

## Test plan

- [x] `pytest tests/test_triton_attention_varlen.py -v` — 8 passed (H100 SXM5)
- [x] `pytest tests/test_attention.py -v` — 26 passed, no regressions
- [x] `ruff check`, `black --check`, `isort --check-only` all clean

## Summary by CodeRabbit

- **New Features**
  - Added optional per-query attention log-sum-exp (LSE) export for dense Triton FlashAttention.
  - Introduced packed variable-length FlashAttention using cumulative sequence lengths, supporting causal, non-causal, decode-style, and zero-length sequences.

- **Documentation**
  - Added operator documentation covering usage, tensor contracts, accuracy references, and limitations.

- **Tests**
  - Added coverage for LSE accuracy, backward behavior, variable-length inputs, and non-contiguous tensors.

- **Benchmarks**
  - Added comparisons with PyTorch baselines for dense and packed attention, including LSE export overhead.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
Thank you for your contribution. Could you add some benchmark tests for the PR content? This will provide some convenience for the reviewers. Thanks!

**Comment by Billy1900** [comment]
## Benchmark results

Added `benchmarks/benchmark_triton_attention_varlen.py` (466be55) covering both additions in this PR:

- `triton_flash_attention(..., return_lse=True/False)` vs. `torch.nn.functional.scaled_dot_product_attention` on dense tensors
- `triton_flash_attention_varlen` vs. a pad-mask-unpad SDPA baseline on packed `cu_seqlens` tensors — the realistic naive alternative for RL rollout batches with skewed per-sequence lengths

Ran on H100 SXM5 (`--iters 10 --warmup 3`, default configs):

**Dense** (causal):

| shape (B×H×S×D) | sdpa fwd ms | triton fwd ms | triton fwd+lse ms | fwd speedup | sdpa f+b ms | triton f+b ms | f+b speedup |
|---|---|---|---|---|---|---|---|
| 2x8x512x64 | 0.018 | 0.060 | 0.071 | 0.30x | 0.655 | 1.816 | 0.36x |
| 2x8x2048x64 | 0.067 | 0.283 | 0.322 | 0.24x | 1.172 | 2.276 | 0.52x |
| 1x16x4096x128 | 0.330 | 0.325 | 0.332 | 1.02x | 1.161 | 2.123 | 0.55x |
| 1x16x8192x128 | 1.009 | 1.113 | 1.107 | 0.91x | 3.683 | 5.586 | 0.66x |

Native SDPA (cuDNN/flash backend) is faster at small/medium shapes and roughly at parity at `head_dim=128`, as expected — this kernel is the cross-platform semantic baseline, not a vendor-tuned kernel. `return_lse=True` adds only ~5-15% over `return_lse=False`, confirming the LSE export is nearly free since `M`/`L` were already computed online.

**Packed varlen** (causal, H=8, D=64):

| config | batch | tokens (pad waste) | pad+sdpa fwd ms | triton varlen fwd ms | fwd speedup | pad+sdpa f+b ms | triton varlen f+b ms | f+b speedup |
|---|---|---|---|---|---|---|---|---|
| uniform_512 | 8 | 4096 (0%) | 1.175 | 0.139 | **8.45x** | 10.054 | 2.496 | 4.03x |
| skewed_short_long | 8 | 10976 (66%) | 3.864 | 0.301 | **12.85x** | 13.853 | 2.647 | 5.23x |
| many_short_few_long | 8 | 4464 (86%) | 4.216 | 0.210 | **20.10x** | 13.460 | 2.631 | 5.12x |
| with_empty_seq | 8 | 10468 (63%) | 3.300 | 0.256 | **12.90x** | 10.879 | 2.685 | 4.05x |

This is where packing pays off — up to 20x forward speedup and several-x memory savings, growing with how skewed the batch's length distribution is. The zero-length-sequence case (`with_empty_seq`) ran cleanly, no NaN/crash.

Reproduce with:
```
python benchmarks/benchmark_triton_attention_varlen.py
```
